### PR TITLE
[ADDED] Force reconnect

### PR DESCRIPTION
**Pull request by piotrpio**
Addresses https://github.com/nats-io/nats-architecture-and-design/issues/259

Signed-off-by: Piotr Piotrowski [piotr@synadia.com](mailto:piotr@synadia.com)

### CODE REVIEW

**Comment by levb** [comment]
@piotrpio `Reconnect()` ignores `opts.AllowReconnect`, is this by design? (asking for the C client)

**Review by Jarema — APPROVED** [review]
LGTM

File: test/reconnect_test.go [hunk 1]
```diff
@@ -884,21 +884,218 @@ func TestAuthExpiredReconnect(t *testing.T) {
...
-			t.Fatalf("Expected to be in reconnecting state after jwt expires, got %v", s)
+	WaitOnChannel(t, stasusCh, nats.RECONNECTING)
+	WaitOnChannel(t, stasusCh, nats.CONNECTED)
+	nc.Close()
+}
+
+func TestForceReconnect(t *testing.T) {
+	s := RunDefaultServer()
+
+	nc, err := nats.Connect(s.ClientURL(), nats.ReconnectWait(10*time.Second))
```
Comment (wallyqs) [inline test/reconnect_test.go:895]: Only one behavior that wanted to confirm is whether we get an error on disconnect or not, and it seems the disconnect callback is called but without an error which I think it is good, e.g.

```go
nats.DisconnectErrHandler(func(_ *nats.Conn, err error) {
	t.Logf("ERROR: %v", err) // error is nil here
}),
```

Comment (piotrpio) [reply test/reconnect_test.go:895]: Yes, the callback is invoked without error which I thought is a reasonable behavior.

**Review by wallyqs — APPROVED** [review]
LGTM

**Pull request merged**